Caching the downloaded archive

ricksbrown opened the pull request:
Allows for caching the downloaded archive - I figured the cache dir might be reused for different OS and arch if it is a network share or mounted in docker or a VM or something, so it handles that scenario.

Also, is it just me or does running `npm install` in this project fail the second time on a Mac...?
So I fixed that too.

Edit: 
To summarise what's in this PR:

* Implements #24 - filesystem caching
* Implements #25 - provide an API to (re-)download binary
* Fixes failure when trying to `npm install` this project more than once on MacOS
* Clean up tmp files when process exits (there was no issue raised for this but it killed my Jenkins agent today because it quickly ran out of disk space - note that filesystem caching alone would alleviate this anyway) - 50 different CI jobs all installing node-chromium creates a LOT of tmp files.

P.S. It's so cool - the install runs so fast the second time around! 😄

ricksbrown commented:
Hey dude - I added another commit to this after looking at #25 - I realised my previous commit would have broken the ability to share the node_modules directory between OSs. I've fixed that now and exposed `install` on the main API to facilitate what they are trying to do:

```js
chromium.install().then(function() {
    // do stuff...
});
```

dtolstyi commented:
Wow. That's amazing, Rick! Thank you. I'm taking a look into it now.

dtolstyi commented on README.MD:
````diff
@@ -40,7 +40,22 @@ npm config set https-proxy http://<username>:<password>@<the.proxy.hostname>:<po
+By default downloaded chromium binaries are not cached on the filesystem. You may enable caching by setting the `CHROMIUM_CACHE` environment variable to a directory path, for example:
+
+```bash
+export CHROMIUM_CACHE=/path/to/cache/rootdir/
````
I'm thinking about the usability for users. Might it be better if we provide a possibility to turn on caching without the need of specifying path?
E.g. having 2 variables:
```
export CHROMIUM_CACHE_ENABLED
export CHROMIUM_CACHE
```
So, when user wants to turn on caching but doesn't want to bother with paths. We can use something default like `.cache` folder where we will save all archives we download. It's even possible to unzip from there instead of tmp path as we do know.

It's more of a usability question and your opinion. For me, it's always much harder to set folder path than to enable a feature. When I'm asked to set a path, I should know if this is absolute / relative path, access is allowed / denied, should I pre-create folder or it will be created by tool, should this folder contain some specific structure, e.t.c.

I'm not fully sure about this feature neither so it would be great if you tell what you think about this.

From what I can see, there might be next use cases for this feature:
1. As described in original request, person wants to have cache at CI system / reuse `node_chromium` folder (though I don't understand how is this possible as most dependencies are actually OS/arch dependant).
2. Testing our own module (much easier to perform checks without the need of installation)
3. When update of module is done (I assume in this case `postinstall` is called as well) - this one looks the most often used and the most important for me

According to **3.** this feature might be very useful if we enable it by default. What do you think about this?

ricksbrown replied:
1. I could see folder reuse happening using docker or perhaps network shares. I replicated it like this:

```bash
docker run -it -v /Users/rick/projects:/projects node /bin/bash
cd /projects/node-chromium
npm run test
````

Then exit docker and run:

```bash
npm install

ls lib/chromium/
# outputs: chrome-linux/ chrome-mac/
```

There may be other scenarios using networked drives or VMs...

2. Yes this makes running the tests much faster and we could potentially avoid the download completely, though that would probably involve putting a chromium archive under source control.

3. YES! You are totally right, this should be on by default. I will rework it to be so.

dtolstyi commented on install.js:
```diff
@@ -26,11 +29,21 @@ function createTempFile() {
+        debug('Found Chromium archive in cache, skipping download');
+        fs.copyFileSync(cacheEntry, tmpPath);
+
+        return Promise.resolve(tmpPath);
```
Great!

dtolstyi commented on install.js:
```diff
@@ -26,11 +29,21 @@ function createTempFile() {
     const url = utils.getDownloadUrl(revision);
 
-    return _downloadFile(url, tmpPath);
+    return _downloadFile(url, tmpPath).then(destPath => {
```
I think it might make sense to rename `desPath` lambda param to `tmpPath` as far as it's actually tmp path.

dtolstyi commented on test/test-cache.js:
```diff
@@ -0,0 +1,115 @@
+    fs.mkdirSync(cacheDir, {recursive: true});
+    process.env.CHROMIUM_CACHE = cacheDir;
+    process.env.npm_config_chromium_cache = cacheDir;
+    return cacheDir;
```
Return seems as not used anywhere.

ricksbrown replied:
👍

dtolstyi commented on test/test-cache.js:
```diff
@@ -0,0 +1,115 @@
+ * Configures node-chromium to use a filesystem cache.
+ */
+function setCacheDir() {
+    delete process.env.CHROMIUM_CACHE;
```
These 2 deletes seems like a good candidate to extract to util function. There's function that gets env variable (from system or .npmrc) and it might be useful to have something like `unsetEnvVariable()` function that will remove both possible variants from system. What do you think?

ricksbrown replied:
Yes!

ricksbrown commented on cache.js:
```diff
@@ -0,0 +1,57 @@
+ * @param {string} revision The Chromium revision to retrieve.
+ * @returns {string} The path to the cached Chromium archive. Falsy if not found.
+ */
+function match(revision) {
```
Yes good point. I think `get` is good and I'll change `put` to `set`.

ricksbrown commented on cache.js:
```diff
@@ -0,0 +1,57 @@
+ * @param {string} revision The Chromium revision in the archive file.
+ * @param {string} file The path to the Chromium archive file to store in cache.
+ */
+function put(revision, file) {
```
ok

ricksbrown commented on cache.js:
```diff
@@ -0,0 +1,57 @@
+ * @param {string} revision The revision of this Chromium binary, essentially a unique cache key.
+ * @returns {string} The cache path, or falsy if caching is not enabled.
+ */
+function getCachePath(revision) {
```
Yes you're right `buildCachePath` is better, doesn't sound like an accessor method.
Regarding the underscore, this method is not included in the `module.exports` which, in CommonJS modules, means it is *actually* private, for real.

dtolstyi commented on README.MD:
```diff
@@ -43,19 +43,33 @@ npm config set no-proxy localhost,127.0.0.1,example.org
-By default downloaded chromium binaries are not cached on the filesystem. You may enable caching by setting the `CHROMIUM_CACHE` environment variable to a directory path, for example:
+By default downloaded chromium binaries are cached in a local directory which varies by operating system:
+
+* Linux: `XDG_CACHE_HOME` or ~/.cache/node-chromium/
```
Are you sure you want to define folders manually for each OS yourself? I think it might be a good idea to hold cache inside: `<your_project>/node_modules/.cache` (so, to have `.cache` folder used by npm). Like this, we don't care about access rights, people having incorrect system variables or smth else, if person can perform `npm install` then we should have all rights to save cache there (actually, 2 of our dependencies save theirs data inside `~/node-modules/.cache` so it looks like a standard practice to do so).

ricksbrown replied:
Yeah fair question - I did a bit of research, there are all sorts of caching specifications and conventions. For example see this discussion: https://github.com/fatiando/pooch/issues/43

I should have mentioned, I basically copied what electron does: https://www.electronjs.org/docs/tutorial/installation#cache

dtolstyi replied:
Yes, I do understand that it might be a solution for big libraries but it requires a lot of support from our side. It means that we are putting each OS and access rights support on ourselves. We need to check that these variables exist, that folders exist, that we got write access there and to do something else otherwise. And, if tomorrow, something changes in this architecture, we will have unexpected failures.
IMO, in our case our cache is a bit different from electron's cache as far as we are caching something related to `npm install` and it actually might reside inside `node_modules` and be shared with it (if someone decides to reuse `node_modules` folder).
What do you think?

ricksbrown replied:
So it's your software and your decision, you can just call it how you want it and I'll make the change.

I think it *looks* more complex than it is.

* every option is in `homedir` of the current user and therefore we are justified in assuming we have write access
* the environment variables`LOCALAPPDATA` and `XDG_CACHE_HOME`  we can assume won't be set accidentally, if they are there we can trust them
* directories are created they don't exist
* if there are errors in caching we log the error to the user but keep going with the installation, so node_chromium will keep functioning regardless and the user can fix whatever issue is logged

So personally I like the idea of following standards for caching. On my machine when I run `ls ~/Library/Caches/` I get a **LOT** of directories listed. When I run `ls ~/.cache` there's about 3 directories.

For my use caching in `node_modules` would be no good because each job on the Jenkins agent has its own node_modules - without a shared cache every CI job will have to download the binary into its own cache using much more disk space and bandwidth.

Obviously it's your baby and you call.

dtolstyi replied:
I see your point, Rick. Actually, you already did so much to this project that I think of it as yours as well, that's why I ask your opinion on what you think is the best. This project is done for the ease of using of everyone, so I don't want to put my own preferences inside but just make it the most useful for everyone.

I'm still thinking of access rights issues, because CI and web servers often got separate user with the least possible rights for execution / access. And, I think that system cache is a bit different purpose than ours module purpose.

I mean, you will always be able to change cache path if you wish but let me think a bit about the better default location.

ricksbrown replied:
Hey I just had a thought - what about we don't "maintain" this problem here?
I could change it to use something like this: https://www.npmjs.com/package/cachedir ?

That way if standards change in future it's just a version bump in the package.json ...?

dtolstyi replied:
@ricksbrown Yes, I absolutely agree that we don't need to do it ourselves and I prefer external package much more. Great, that you found a package that can help here.
Still, the question for me is if we store it in system or in project.
I have to think of possible use cases (multiple node installations using `nvm` / multiple projects on the same system / CI servers / ...). 

Thank you for your help, Rick. gn!

dtolstyi commented on install.js:
```diff
@@ -90,10 +93,10 @@ async function install() {
 
         console.info(`Step 2. Downloading Chromium (this might take a while). Revision number: ${revision}`);
-        const tmpPath = await downloadChromiumRevision(revision);
+        const archivePath = await downloadChromiumRevision(revision);
```
Nice!

dtolstyi commented on test/_utils.js:
```diff
@@ -1,10 +1,26 @@
+ * @param {string} name The UPPER_CASE name of the environment variable.
+ * @param {string} value The value to set - if falsy will be deleted.
+ */
+function setEnvVar(name, value) {
```
This one I don't like very much. I prefer functions to perform only single action instead of being multi-purpose and performing different tasks according to theirs params. I would prefer it to be split into 2 params.

In addition, there's a possible bug: imagine `0` or `false` passed as `value`. In this case it will just delete env vars instead of setting theirs values.

dtolstyi replied:
Oh, I just noticed that this is test-only but still, for readability, I would prefer to have `set` and `delete`. Like this, when you read test's code you can see that intention was actually to delete value and not just set it empty or something else.

ricksbrown replied:
Yes I think you are right about that.

dtolstyi commented:
Wow, I just checked Progress Bar and it looks really amazing! Thank you Rick!

ricksbrown commented:
Ok yep - I've removed the progress bar from this PR.

dtolstyi commented on README.MD:
````diff
@@ -40,7 +40,36 @@ npm config set https-proxy http://<username>:<password>@<the.proxy.hostname>:<po
+You may override the cache path by setting the `CHROMIUM_CACHE` environment variable to a directory path, for example:
+
+```bash
+export CHROMIUM_CACHE=/path/to/cache/dir/
````
I think it might be a good idea to change env var name to specifically `NODE_CHROMIUM_CACHE` because it might get mixed with Chromium env variables itself.

dtolstyi commented on README.MD:
````diff
@@ -40,7 +40,36 @@ npm config set https-proxy http://<username>:<password>@<the.proxy.hostname>:<po
+You may disable caching by setting `CHROMIUM_CACHE_SKIP` to `true`:
+
+```bash
+export CHROMIUM_CACHE_SKIP=true
````
And the same here with the name of env var. I would suggest to replace `SKIP` with `DISABLE`. Because skip looks for me like we are still saving everything into cache but we are just skipping it when downloading/unzipping. And disable looks more like we are actually disabling the cache.
I like something like: `DISABLE_NODE_CHROMIUM_CACHE`.

ricksbrown commented:
> OK, looks like we got 2 points remaining to solve and we are ready to go:
> 
> 1. Env Vars naming
> 2. `.cache` folder location

Thanks for all your feedback on the PR, your points have been wise and constructive, you sure know your stuff. It's amazing how much better people can make things when they work together.

I think the cache module looks so much better now using [node-cachedir](https://www.npmjs.com/package/cachedir) - I definitely had that wrong first time around trying to manage it ourselves. I think it's in safe hands now, that module gets well over 2 million downloads from NPM a week, so it is mature and well-tested. 

I'm really happy with it now, but if you still want to change it that's your call...

dtolstyi commented:
Thank you for all your work, suggestions and kind words Rick. I do really appreciate it. 
I still think that all the merit here goes to you as far as I haven't done much and all was done by you.

I'm almost convinced to use `cachedir` module you find which is so widely-used. 

It's quite late here, so I don't want to merge and release a new version now (merging/releasing late night usually leads to very bad consequences). And, I would like to add **Contributors** section to README before release. So, I will try to do it tomorrow morning.

I need your advice regarding the version: do you think this is major release? For me it looks like one (I'm thinking about version 3.0.0). Here, it's quite complicated to tell but looks like back compatibility is broken in this version as far as we introduced caching by default what was not the case before. So, I'm not sure if this is actually 3.0.0 or 2.4.0. What do you think?

And could you open that PR with progress bar, please? I would love to have it available in next version as well.

Thank you once more for all your help! It's really huge and I do appreciate it!

ricksbrown commented:
Yes on not making decisions late at night! I hear you man 🦉 

Hmmm, regarding the version bump - strictly speaking the **API** is backwards compatible with the previous release so it would probably be a minor increment (not 3.0.0). 

I see the caching as an implementation detail (which can easily be disabled to restore the old behavior).
As a consumer when I see a major version increment I assume things are likely to break - if I pick them up I know I might have to change my own code too - this PR is not likely to break anybody's existing code. 

I was going to ask if you think we should rename these for consistency:

`CHROMIUM_DOWNLOAD_HOST` > `NODE_CHROMIUM_DOWNLOAD_HOST`
`CHROMIUM_REVISION` > `NODE_CHROMIUM_REVISION`

If so, that would be a definite major version bump to 3.0.0 because it would not be backwards compatible.

FWIW I think the previous release might have been more of a candidate for a major version bump since we modified the export of install.js and changed the supported NodeJS version. For example the code honeo posted in #25 will no longer work in the previous release (they aren't using your official documented API though). 

Perhaps this PR and the previous one combined do add up to a 3.0.0

Ultimately you don't need to stick to strict semver anyway. I've seen people bump the major version just to indicate significant changes and I think that could apply here. 

I'd go with your gut feeling.

dtolstyi commented:
Hi Rick,

Thank you for your recommendations.

Regarding the variables names, I've already been thinking about this and I got some mixed feelings.
From one side, I do love when everything is of the same format (all our variables start from `NODE_CHROMIUM`).
From the other side, it doesn't seem very intuitive and nice the name: `NODE_CHROMIUM_REVISION` as far as it pretends to specify revision of node_chromium and not chromium itself. But, if we look at it as a parameter with just `NODE_CHROMIUM` prefix is starts to make sense. 

Not sure if I managed to explain my feelings correctly but they are somehow mixed. 

But, if you mentioned this as well, I think it would be a good idea to have all our variables prefixed as `NODE_CHROMIUM`. I prefer this. Could you change old variables to a new format and then we will release a new major version?

dtolstyi commented:
Great! Thank you, Rick!

dtolstyi commented on README.MD:
````diff
@@ -121,7 +121,27 @@ async function takeScreenshot(driver, name) {
+# node_chromium_cache=/path/to/cache/dir/
+```
+
+You may disable caching by setting `NODE_CHROMIUM_CACHE_DISABLE` to `true`:
````
Last request regarding this variable, Rick: do you mind if we change it to: `NODE_CHROMIUM_DISABLE_CACHE`? For me it sounds more natural like this. What do you think?

ricksbrown replied:
I can see both perspectives: 

* consistency (everything cache related starts with `NODE_CHROMIUM_CACHE` 
vs
* English verb - object word order.

I really don't mind, both are legit. Want me to change it?

dtolstyi replied:
Makes sense what you say but I would like then `NODE_CHROMIUM_CACHE` variable to be renamed to `NODE_CHROMIUM_CACHE_PATH`.
Like this, it looks natural for me:
```
NODE_CHROMIUM_CACHE_PATH
NODE_CHROMIUM_CACHE_DISABLE
```
Then I'm absolutely fine with all.

ricksbrown replied:
Bingo, that's a winner right there!

dtolstyi replied:
Could you do this, Rick, or you want me to do it?

ricksbrown commented:
Hey man, I just tried the progressbar against the taobao mirror - the stream does not have the necessary HTTP headers.
Hope you don't mind, I'm going to throw in a fix for that too...

dtolstyi commented:
Sure, no problem))

dtolstyi commented:
Wow. I thought that this PR is going to be very easy but you made so many great decisions and implementations in addition to initial plan. Thank you so much for all your efforts @ricksbrown !

ricksbrown commented:
> Wow. I thought that this PR is going to be very easy but you made so many great decisions and implementations in addition to initial plan. Thank you so much for all your efforts @ricksbrown !

Yeah me too... I thought it was just about ready to go after my first commit - haha! Your feedback took it to the next level. Thanks for that. I look forward to using this for real now.
